README.md
# scale_button

A customizable and easy-to-use animation button library for Flutter. Compatible with Android & iOS & Web. :smiley:

[![pub](https://img.shields.io/pub/v/scale_button)](https://pub.dev/packages/scale_button)



<br>

## Showcase

<img src = "https://user-images.githubusercontent.com/55150540/145807932-1500e1ad-745c-4ac5-bf9e-588a4760b414.gif" width = 200>

<br>

## Getting Started

In the pubspec.yaml of your flutter project, add the following dependency:
```yaml
dependencies:
  ...
  scale_button : "^lastest_version"
```

<br>

## Usage
Need to include the import the package to the dart file where it will be used, refer the below command
```dart
import 'package:scale_button/scale_button.dart';
```

<br>

## Complete example
```dart
import 'package:flutter/material.dart';

import 'package:scale_button/scale_button.dart';

void main() {
  runApp(MaterialApp(home: const MyApp()));
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text("Scale Button Example"),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          ScaleButton(
            child: Container(
              height: 48.0,
              width: 300.0,
              alignment: Alignment.center,
              decoration: const BoxDecoration(
                color: Colors.blueAccent,
                borderRadius: BorderRadius.all(Radius.circular(24.0)),
              ),
              child: Text(
                "Basic",
                style: TextStyle(color: Colors.white),
              ),
            ),
          ),
          const SizedBox(height: 14.0),
          ScaleButton(
            reverse: true,
            child: Container(
              height: 48.0,
              width: 300.0,
              alignment: Alignment.center,
              decoration: const BoxDecoration(
                color: Colors.blueAccent,
                borderRadius: BorderRadius.all(Radius.circular(24.0)),
              ),
              child: Text(
                "Basic (reverse)",
                style: TextStyle(color: Colors.white),
              ),
            ),
          ),
          const SizedBox(height: 36.0),
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const ScaleButton(
                bound: 0.5,
                duration: Duration(milliseconds: 1500),
                child: Text("Text Button"),
              ),
              ScaleButton(
                bound: 0.8,
                reverse: true,
                child: Container(
                  height: 100.0,
                  width: 150.0,
                  decoration: const BoxDecoration(
                      color: Colors.orangeAccent, shape: BoxShape.circle),
                  alignment: Alignment.center,
                  child: Text(
                    "Custom Button\n(reverse)",
                    style: TextStyle(
                      color: Colors.white,
                      fontSize: 11.0,
                    ),
                    textAlign: TextAlign.center,
                  ),
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }
}

```


CHANGELOG.md
## [2.1.2] - 2021-12-13

* Update README


## [2.1.1] - 2021-12-13

* Update README


## [2.1.0] - 2021-12-11

* Update README
* Remove 'width' parameter
* Remove 'height' parameter
* Remove 'boxDecoration' parameter


## [2.0.0] - 2021-03-11

* Stable null safety release.


## [1.1.1] - 2020-12-3

* Update README



## [1.1.0] - 2020-12-2

* Refactoring code


## [1.0.1] - 2020-12-2

* Update pubspec.yaml
* Update README

## [1.0.0] - 2020-12-2

* Scale Button creation



lib/scale_button.dart
import 'package:flutter/material.dart';

class ScaleButton extends StatefulWidget {
  const ScaleButton({
    Key? key,
    Duration duration = const Duration(milliseconds: 300),
    double bound = 0.2,
    VoidCallback? onTap,
    bool reverse = false,
    required Widget child,
  })  : this.duration = duration,
        this.bound = bound,
        this.onTap = onTap,
        this.child = child,
        this.reverse = reverse,
        super(key: key);

  /// Adjust the animation speed
  ///
  /// Defaults to [300]
  final Duration duration;

  /// Adjust the shrink & zoom size.
  ///
  /// Default to [0.2]
  final double bound;

  /// Called when the user taps this part of the material.
  final VoidCallback? onTap;

  /// The widget below this widget in the tree.
  final Widget child;

  /// You can choose the type of animation [true : shrink, false: zoom]
  ///
  /// Default to [false]
  final bool reverse;

  @override
  _ScaleButtonState createState() => _ScaleButtonState();
}

class _ScaleButtonState extends State<ScaleButton>
    with SingleTickerProviderStateMixin {
  double _scale = 1.0;
  late AnimationController _controller;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      vsync: this,
      duration: widget.duration,
      upperBound: widget.bound,
    )..addListener(() {
        setState(() {
          _scale =
              widget.reverse ? 1 + _controller.value : 1 - _controller.value;
        });
      });
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTapDown: (_) {
        _controller.forward();
      },
      onTapUp: (_) {
        _controller.reverse();
      },
      onTapCancel: () {
        _controller.reverse();
      },
      onTap: widget.onTap,
      child: Transform.scale(
        scale: _scale,
        child: widget.child,
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:scale_button/scale_button.dart';

void main() {
  runApp(MaterialApp(home: const MyApp()));
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text("Scale Button Example"),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          ScaleButton(
            child: Container(
              height: 48.0,
              width: 300.0,
              alignment: Alignment.center,
              decoration: const BoxDecoration(
                color: Colors.blueAccent,
                borderRadius: BorderRadius.all(Radius.circular(24.0)),
              ),
              child: Text(
                "Basic",
                style: TextStyle(color: Colors.white),
              ),
            ),
          ),
          const SizedBox(height: 14.0),
          ScaleButton(
            reverse: true,
            child: Container(
              height: 48.0,
              width: 300.0,
              alignment: Alignment.center,
              decoration: const BoxDecoration(
                color: Colors.blueAccent,
                borderRadius: BorderRadius.all(Radius.circular(24.0)),
              ),
              child: Text(
                "Basic (reverse)",
                style: TextStyle(color: Colors.white),
              ),
            ),
          ),
          const SizedBox(height: 36.0),
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const ScaleButton(
                bound: 0.5,
                duration: Duration(milliseconds: 1500),
                child: Text("Text Button"),
              ),
              ScaleButton(
                bound: 0.8,
                reverse: true,
                child: Container(
                  height: 100.0,
                  width: 150.0,
                  decoration: const BoxDecoration(
                      color: Colors.orangeAccent, shape: BoxShape.circle),
                  alignment: Alignment.center,
                  child: Text(
                    "Custom Button\n(reverse)",
                    style: TextStyle(
                      color: Colors.white,
                      fontSize: 11.0,
                    ),
                    textAlign: TextAlign.center,
                  ),
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }
}
